pycbc/waveform/waveform_modes.py

```diff
@@ -0,0 +1,491 @@
+    out = None
+    for mode in hlms:
+        ell, m = mode
+        hlm = hlms[ell, m]
+        ylm = lal.SpinWeightedSphericalHarmonic(inclination, phi, -2, ell, m)
+        if out is None:
+            out = ylm * hlm
+        else:
+            out += ylm * hlm
+    return out
+
+
+def default_modes(approximant):
+    """Returns the default modes for the given approximant.
+    """
+    # FIXME: this should be replaced to a call to a lalsimulation function,
+    # whenever that's added
+    if approximant in ['IMRPhenomXPHM', 'IMRPhenomXHM']:
+        # according to arXiv:2004.06503
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant in ['IMRPhenomPv3HM', 'IMRPhenomHM']:
+        # according to arXiv:1911.06050
+        ma = [(2, 2), (2, 1), (3, 3), (3, 2), (4, 4), (4, 3)]
+        # add the -m modes
+        ma += [(ell, -m) for ell, m in ma]
+    elif approximant.startswith('NRSur7dq4'):
+        # according to arXiv:1905.09300
+        ma = [(ell, m) for ell in [2, 3, 4] for m in range(-ell, ell+1)]
+    else:
+        raise ValueError("I don't know what the default modes are for "
+                         "approximant {}, sorry!".format(approximant))
+    return ma
+
+
+def get_glm(ell, m, theta):
+    """The inclination part of the ylm."""
+    return lal.SpinWeightedSphericalHarmonic(theta, 0., -2, ell, m).real
+
+
+def get_nrsur_modes(**params):
+    """Generates NRSurrogate waveform mode-by-mode.
+
+    All waveform parameters should be provided as keyword arguments.
+    Recognized parameters are listed below. Unrecognized arguments are ignored.
+
+    Parameters
+    ----------
+    template: object
+        An object that has attached properties. This can be used to substitute
+        for keyword arguments. A common example would be a row in an xml table.
+    approximant : str
+        The approximant to generate. Must be one of the ``NRSur*`` models.
+    {delta_t}
+    {mass1}
+    {mass2}
+    {spin1x}
+    {spin1y}
+    {spin1z}
+    {spin2x}
+    {spin2y}
+    {spin2z}
+    {f_lower}
+    {f_ref}
+    {distance}
+    {mode_array}
+
+    Returns
+    -------
+    dict :
+        Dictionary of ``(l, m)`` -> ``(h_+, -h_x)`` ``TimeSeries``.
+    """
+    laldict = _check_lal_pars(params)
+    ret = lalsimulation.SimInspiralPrecessingNRSurModes(
+        params['delta_t'],
+        params['mass1']*lal.MSUN_SI,
+        params['mass2']*lal.MSUN_SI,
+        params['spin1x'], params['spin1y'], params['spin1z'],
+        params['spin2x'], params['spin2y'], params['spin2z'],
+        params['f_lower'], params['f_ref'],
+        params['distance']*1e6*lal.PC_SI, laldict,
+        getattr(lalsimulation, params['approximant'])
+    )
+    hlms = {}
+    while ret:
+        hlm = TimeSeries(ret.mode.data.data, delta_t=ret.mode.deltaT,
+                         epoch=ret.mode.epoch)
+        hlms[ret.l, ret.m] = (hlm.real(), hlm.imag())
```

Comment:
Just to check, does this have to copy the data in memory, or does it just change the pointers around? If the former, is the cost of this negligble?

Reply:
Good question... I'm not sure. I remember being burned by not copying data returned by lalsimulation in the past, but that was several years ago; not sure if it's changed or not. Looking at what's done in `get_td_waveform` though (https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/waveform.py#L214) the data is copied, so probably best to let it copy.